android/app/src/main/kotlin/com/nt4f04und/sweyer/MusicPlayerAppWidget.kt

```diff
@@ -0,0 +1,319 @@
+                if (isPortraitOrientation) { // Depends on the home-screen orientation
+                    width = mAppWidgetOptions.getInt(AppWidgetManager.OPTION_APPWIDGET_MIN_WIDTH).dipToPixels()
+                    height = mAppWidgetOptions.getInt(AppWidgetManager.OPTION_APPWIDGET_MAX_HEIGHT).dipToPixels()
+                } else {
+                    width = mAppWidgetOptions.getInt(AppWidgetManager.OPTION_APPWIDGET_MAX_WIDTH).dipToPixels()
+                    height = mAppWidgetOptions.getInt(AppWidgetManager.OPTION_APPWIDGET_MIN_HEIGHT).dipToPixels()
+                }
+            } else {
+                // Get min dimensions from provider info
+                val providerInfo: AppWidgetProviderInfo? = appWidgetManager.getAppWidgetInfo(appWidgetId)
+                if (providerInfo == null) {
+                    val displayMetrics = Resources.getSystem().displayMetrics
+                    width = displayMetrics.widthPixels
+                    height = displayMetrics.heightPixels
+                } else {
+                    width = providerInfo.minWidth.dipToPixels()
+                    height = providerInfo.minHeight.dipToPixels()
+                }
+            }
+            if (maxWidth < width) {
+                maxWidth = width
+            }
+            if (maxHeight < height) {
+                maxHeight = height
+            }
+            Size(width, height)
+        }.toTypedArray()
+        return SizeInfo(sizes, Size(maxWidth, maxHeight))
+    }
+
+    /**
+     * Update the appearance of each app widget by calling the [builder] with the size of the corresponding widget.
+     */
+    fun updateAll(builder: (Size) -> RemoteViews) {
+        for ((appWidgetId, size) in appWidgetIds.zip(sizeInfo.sizes)) {
+            appWidgetManager.updateAppWidget(appWidgetId, builder(size))
+        }
+    }
+}
+
+/**
+ * Utility to build background intents to interact with the AudioService from the widget.
+ */
+object AudioServiceBackgroundIntent {
+    /**
+     * Build a pending intent for the given [context] emulating a key-press of the given [keyEvent] to the AudioService.
+     */
+    private fun getIntent(context: Context, keyEvent: KeyEvent): PendingIntent {
+        val intent = Intent(context, com.ryanheise.audioservice.AudioService::class.java)
+        intent.action = Intent.ACTION_MEDIA_BUTTON
+        intent.putExtra(Intent.EXTRA_KEY_EVENT, keyEvent)
+        var flags = PendingIntent.FLAG_UPDATE_CURRENT
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+            flags = flags or PendingIntent.FLAG_IMMUTABLE
+        }
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
+            intent.identifier = keyEvent.toString()
+        } else {
+            intent.type = keyEvent.toString()
+        }
+        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+            PendingIntent.getForegroundService(context, 0, intent, flags)
```

Comment:
I'm looking into the ANR, but we definitely need to use `getForegroundService`, because otherwise the intent will be ignored on newer Android versions.

Reply:
Looks like we might be able to work around this issue by using the broadcast receiver that listens to media playback button presses and that starts the service for us.

During my testing this worked very well, except for one time where we failed to create a notification with an `ForegroundServiceStartNotAllowedException`. This was after killing the app and I think it was caused by an extremely slow startup, which might have caused us to exceed some timeframe. But I have not been able to replicate that, it worked every other attempt.